safe-global/safe-wallet-web
# refactor: switch to Yarn v4

## What it solves
The first step to switching to a monorepo structure is to update yarn as newer versions offer:
- improved speed
- better monorepo management

We were stuck with yarn classic (which is deprecated for more than a couple of years now). Yarn 4 feels significantly faster and will lead to a better developer experience

## How this PR fixes it
I've added a packagerManager field to the yarn.lock file. Had to dedupe some dependencies as they were wrong in yarn v1 and also had to remove `@web3-onboard/keystone` as it has one incorrectly defined dependency (the dependency version is not following semantic versioning and I couldn't find a way for yarn to ignore this). I have created a fix for it https://github.com/blocknative/web3-onboard/pull/2322, but I doubt that it will be merged any time soon. 

About github actions. 
Our github actions were incorrectly placed in the workflows folder. I moved them to a separate github actions folder. This also made it possible to use tools such as act https://github.com/nektos/act to tests the flows locally. 

The caching dependency action is a bit weird, but I had my reasons :). Using the `actions/cache@v4` directly turned out to be very slow. I think that the main issue is the use of hashFiles in the key. For example restoring next.js build cache the key there is composed of `hashFiles('**/*.js', '**/*.jsx', '**/*.ts', '**/*.tsx')` - we are dealing with a lot of files here and that apparently takes a lot of time. That's why I'm caching the key once it is generated by the restore action and then using it for the save action.
 
## How to test it
Checkout the branch, make sure that you have run `corepack enable` and then just run `yarn` in the terminal. This should install the latest yarn v4 version and all necessary project dependencies.

## Screenshots

## Checklist
